helper/plugin/grpc_provider.go

```diff
@@ -0,0 +1,616 @@
+
+	res := s.provider.ResourcesMap[req.TypeName]
+	block := res.CoreConfigSchema()
+
+	stateVal, err := msgpack.Unmarshal(req.CurrentState.Msgpack, block.ImpliedType())
+	if err != nil {
+		resp.Diagnostics = appendDiag(resp.Diagnostics, err)
+		return resp, nil
+	}
+
+	instanceState := schema.InstanceStateFromStateValue(stateVal, res.SchemaVersion)
+
+	newInstanceState, err := res.RefreshWithoutUpgrade(instanceState, s.provider.Meta())
+	if err != nil {
+		resp.Diagnostics = appendDiag(resp.Diagnostics, err)
+		return resp, nil
+	}
+
+	// helper/schema should always copy the ID over, but do it again just to be safe
+	newInstanceState.Attributes["id"] = newInstanceState.ID
+
+	newConfigVal, err := hcl2shim.HCL2ValueFromFlatmap(newInstanceState.Attributes, block.ImpliedType())
+	if err != nil {
+		resp.Diagnostics = appendDiag(resp.Diagnostics, err)
+		return resp, nil
+	}
+
+	newConfigMP, err := msgpack.Marshal(newConfigVal, block.ImpliedType())
+	if err != nil {
+		resp.Diagnostics = appendDiag(resp.Diagnostics, err)
+		return resp, nil
+	}
+
+	resp.NewState = &proto.DynamicValue{
+		Msgpack: newConfigMP,
+	}
+
+	return resp, nil
+}
+
+func (s *GRPCProviderServer) PlanResourceChange(_ context.Context, req *proto.PlanResourceChange_Request) (*proto.PlanResourceChange_Response, error) {
+	resp := &proto.PlanResourceChange_Response{}
+
+	res := s.provider.ResourcesMap[req.TypeName]
+	block := res.CoreConfigSchema()
+
+	priorStateVal, err := msgpack.Unmarshal(req.PriorState.Msgpack, block.ImpliedType())
+	if err != nil {
+		resp.Diagnostics = appendDiag(resp.Diagnostics, err)
+		return resp, nil
+	}
+
+	proposedNewStateVal, err := msgpack.Unmarshal(req.ProposedNewState.Msgpack, block.ImpliedType())
+	if err != nil {
+		resp.Diagnostics = appendDiag(resp.Diagnostics, err)
+		return resp, nil
+	}
+
+	info := &terraform.InstanceInfo{
+		Type: req.TypeName,
+	}
+
+	priorState := schema.InstanceStateFromStateValue(priorStateVal, res.SchemaVersion)
+
+	// turn the propsed state into a legacy configuration
+	config := terraform.NewResourceConfigShimmed(proposedNewStateVal, block)
+
+	diff, err := s.provider.Diff(info, priorState, config)
+	if err != nil {
+		resp.Diagnostics = appendDiag(resp.Diagnostics, err)
+		return resp, nil
+	}
+
+	// now we need to apply the diff to the prior state, so get the planned state
+	plannedStateVal, err := schema.ApplyDiff(priorStateVal, diff, block)
+	if err != nil {
+		resp.Diagnostics = appendDiag(resp.Diagnostics, err)
+		return resp, nil
+	}
+
+	plannedMP, err := msgpack.Marshal(plannedStateVal, block.ImpliedType())
+	if err != nil {
+		resp.Diagnostics = appendDiag(resp.Diagnostics, err)
+		return resp, nil
+	}
+	resp.PlannedState.Msgpack = plannedMP
+
+	// the Meta field gets encoded into PlannedPrivate
+	plannedPrivate, err := json.Marshal(diff.Meta)
+	if err != nil {
+		resp.Diagnostics = appendDiag(resp.Diagnostics, err)
+		return resp, nil
+	}
+	resp.PlannedPrivate = plannedPrivate
```

Comment:
This reminds me that over on the Core side of this we still have `Private` declared as `cty.Value`. No action required here, but when we start wiring this in we'll need to change that to `[]byte` first and update the file read/write/upgrade code to expect this.